src/Symfony/Bundle/FrameworkBundle/Test/TestContainer.php

```diff
@@ -11,95 +11,138 @@
+        $this->kernel = $kernel;
+        $this->privateServicesLocatorId = $privateServicesLocatorId;
     }
 
     /**
      * {@inheritdoc}
      */
     public function compile()
     {
-        $this->publicContainer->compile();
+        $this->getPublicContainer()->compile();
     }
 
     /**
      * {@inheritdoc}
      */
     public function isCompiled()
     {
-        return $this->publicContainer->isCompiled();
+        return $this->getPublicContainer()->isCompiled();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getParameterBag()
+    {
+        return $this->getPublicContainer()->getParameterBag();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getParameter($name)
+    {
+        return $this->getPublicContainer()->getParameter($name);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasParameter($name)
+    {
+        return $this->getPublicContainer()->hasParameter($name);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setParameter($name, $value)
+    {
+        $this->getPublicContainer()->setParameter($name, $value);
     }
 
     /**
      * {@inheritdoc}
      */
     public function set($id, $service)
     {
-        $this->publicContainer->set($id, $service);
+        $this->getPublicContainer()->set($id, $service);
     }
 
     /**
      * {@inheritdoc}
      */
     public function has($id)
     {
-        return $this->publicContainer->has($id) || $this->privateContainer->has($id);
+        return $this->getPublicContainer()->has($id) || $this->getPrivateContainer()->has($id);
     }
 
     /**
      * {@inheritdoc}
      */
     public function get($id, $invalidBehavior = /* self::EXCEPTION_ON_INVALID_REFERENCE */ 1)
     {
-        return $this->privateContainer->has($id) ? $this->privateContainer->get($id) : $this->publicContainer->get($id, $invalidBehavior);
+        return $this->getPrivateContainer()->has($id) ? $this->getPrivateContainer()->get($id) : $this->getPublicContainer()->get($id, $invalidBehavior);
     }
 
     /**
      * {@inheritdoc}
      */
     public function initialized($id)
     {
-        return $this->publicContainer->initialized($id);
+        return $this->getPublicContainer()->initialized($id);
     }
 
     /**
      * {@inheritdoc}
      */
     public function reset()
     {
-        $this->publicContainer->reset();
+        $this->getPublicContainer()->reset();
     }
 
     /**
      * {@inheritdoc}
      */
     public function getServiceIds()
     {
-        return $this->publicContainer->getServiceIds();
+        return $this->getPublicContainer()->getServiceIds();
     }
 
     /**
      * {@inheritdoc}
      */
     public function getRemovedIds()
     {
-        return $this->publicContainer->getRemovedIds();
+        return $this->getPublicContainer()->getRemovedIds();
+    }
+
+    private function getPublicContainer()
+    {
+        $this->kernel->boot();
```

Comment:
Haven't we made a change in 4.1 implying that booting an already booted kernel is not a no-op ? That's would be a bad side-effect then here.